src/whitenoise/users.rs

```diff
@@ -35,14 +36,81 @@ impl User {
     /// * `whitenoise` - The Whitenoise instance used to access the Nostr client and database
     pub async fn sync_metadata(&mut self, whitenoise: &Whitenoise) -> Result<()> {
         let relays_to_query = self.get_query_relays(whitenoise).await?;
-        let metadata = whitenoise
+        let metadata_result = whitenoise
             .nostr
             .fetch_metadata_from(&relays_to_query, self.pubkey)
             .await?;
-        if let Some(metadata) = metadata {
+        if let Some((metadata, event_timestamp, event_id)) = metadata_result {
             if self.metadata != metadata {
-                self.metadata = metadata;
-                self.save(&whitenoise.database).await?;
+                let event_datetime =
+                    DateTime::from_timestamp_millis((event_timestamp.as_u64() * 1000) as i64)
+                        .ok_or_else(|| {
+                            WhitenoiseError::Database(
+                                crate::whitenoise::database::DatabaseError::InvalidTimestamp {
+                                    timestamp: (event_timestamp.as_u64() * 1000) as i64,
+                                },
+                            )
+                        })?;
+
+                // Check if this event is newer than our most recent processed metadata event
+                let newest_processed_timestamp = ProcessedEvent::newest_event_timestamp_for_kinds(
+                    None, // Global events (user metadata)
+                    &[0], // Metadata events are kind 0
+                    &whitenoise.database,
+                )
+                .await
+                .map_err(WhitenoiseError::Database)?;
+
+                let should_update = match newest_processed_timestamp {
+                    None => {
+                        tracing::debug!(
+                            target: "whitenoise::users::sync_metadata",
+                            "No processed metadata events for user {}, accepting new event",
+                            self.pubkey
+                        );
+                        true
+                    }
+                    Some(stored_timestamp) => {
+                        let is_newer_or_equal = event_datetime.timestamp_millis()
+                            >= stored_timestamp.timestamp_millis();
+                        if !is_newer_or_equal {
+                            tracing::debug!(
+                                target: "whitenoise::users::sync_metadata",
+                                "Ignoring stale metadata event for user {} (event: {}, stored: {})",
+                                self.pubkey,
+                                event_datetime.timestamp_millis(),
+                                stored_timestamp.timestamp_millis()
+                            );
+                        }
+                        is_newer_or_equal
+                    }
+                };
```

Comment:
Can we maybe move all of this to a private `should_update_metadata` or similar private method?